allennlp/training/trainer.py

```diff
@@ -631,6 +631,9 @@ def train(self) -> None:
 
         logger.info("Beginning training.")
 
+        train_metrics = None
+        val_metrics = None
```

Comment:
if you use empty dicts as the default values, then you don't have to check for `None` below, iterating over the items will just be a no-op

Reply:
@joelgrus I had this originally.  Unfortunately `train_metrics` and `val_metrics` can be updated to `None` in the loop.  Do you think I should address this in the upstream methods?  It does seem like undesirable (and confusing) behavior...

Reply:
I don't see where `train_metrics` can be set to `None`, and it looks like if `val_metrics` were initialized as an empty dict you could remove the code that sets it to `None`

Reply:
https://github.com/schmmd/allennlp/blob/ce154e82c4ed3c2adf07a70fdb386f4f131b6837/allennlp/training/trainer.py#L662

I'll change it to empty dict.